Backfill_Correction/delphiBackfillCorrection/R/model.R

```diff
@@ -0,0 +1,259 @@
+
+          coefs_result[[success+1]] = coef(obj)
+        }
+
+        success = success + 1
+      },
+      error=function(e) {print(paste("Training failed for", model_path, sep=" "))}
+    )
+  }
+  if (success < 9) {return (NULL)}
+  if (!make_predictions) {return (list())}
+  
+  coef_combined_result = data.frame(tau=taus, geo=geo, test_lag=test_lag)
+  coef_combined_result[coef_list] = as.matrix(do.call(rbind, coefs_result))
+  
+  return (list(test_data, coef_combined_result))
+}
+
+#' Evaluation of the test results based on WIS score
+#' The WIS score calculation is based on the weighted_interval_score function 
+#' from the `evalcast` package from Delphi
+#' 
+#' @param test_data dataframe with a column containing the prediction results of
+#'    each requested quantile. Each row represents an update with certain
+#'    (reference_date, issue_date, location) combination.
+#' @template taus-template
+#' 
+#' @importFrom evalcast weighted_interval_score
+#' 
+#' @export
+evaluate <- function(test_data, taus) {
+  n_row = nrow(test_data)
+  taus_list = as.list(data.frame(matrix(replicate(n_row, taus), ncol=n_row)))
+  
+  # Calculate WIS
+  predicted_all = as.matrix(test_data[c("predicted_tau0.01", "predicted_tau0.025",
+                                        "predicted_tau0.1", "predicted_tau0.25",
+                                        "predicted_tau0.5", "predicted_tau0.75",
+                                        "predicted_tau0.9", "predicted_tau0.975",
+                                        "predicted_tau0.99")])
+  predicted_all_exp = exp(predicted_all)
+  predicted_trans = as.list(data.frame(t(predicted_all - test_data$log_value_target)))
+  test_data$wis = mapply(weighted_interval_score, taus_list, predicted_trans, 0)
+  
+  return (test_data)
+}
+
+#' Train model using quantile regression with Lasso penalty, or load from disk
+#'
+#' @param model_path path to read model from or to save model to
+#' @template train_data-template
+#' @template covariates-template
+#' @param tau decimal quantile to be predicted. Values must be between 0 and 1.
+#' @template lp_solver-template
+#' @template lambda-template
+#' @template train_models-template
+#'
+#' @importFrom quantgen quantile_lasso
+#' @importFrom stringr str_interp
+get_model <- function(model_path, train_data, covariates, tau,
+            lambda, lp_solver, train_models) {
+  if (train_models || !file.exists(model_path)) {
+    if (!train_models && !file.exists(model_path)) {
+      warning(str_interp("user requested use of cached model but file {model_path}"),
+        " does not exist; training new model")
+    }
+    # Quantile regression
+    obj <- quantile_lasso(as.matrix(train_data[covariates]),
+                         train_data$log_value_target, tau = tau,
+                         lambda = lambda, standardize = FALSE, lp_solver = lp_solver)
+
+    # Save model to cache.
+    create_dir_not_exist(dirname(model_path))
+    save(obj, file=model_path)
```

Comment:
this might be a better place to record the date trained for this model

Reply:
We do have the `training_end_date` serving as the prefix of the saved model names.

Reply:
Addressed in https://github.com/cmu-delphi/covidcast-indicators/pull/1711